frontend/src/components/common/Table/Table.tsx

```diff
@@ -0,0 +1,288 @@
+   * Function to filter the rows
+   * Works in addition to the default table filtering and searching
+   */
+  filterFunction?: (item: RowItem) => boolean;
+  /**
+   * Whether to show a loading spinner
+   */
+  loading?: boolean;
+};
+
+// Use a zero-indexed "useURLState" hook, so pages are shown in the URL as 1-indexed
+// but internally are 0-indexed.
+function usePageURLState(
+  key: string,
+  prefix: string,
+  initialPage: number
+): ReturnType<typeof useURLState> {
+  const [page, setPage] = useURLState(key, { defaultValue: initialPage + 1, prefix });
+  const [zeroIndexPage, setZeroIndexPage] = useState(page - 1);
+
+  useEffect(() => {
+    setZeroIndexPage((zeroIndexPage: number) => {
+      if (page - 1 !== zeroIndexPage) {
+        return page - 1;
+      }
+
+      return zeroIndexPage;
+    });
+  }, [page]);
+
+  useEffect(() => {
+    setPage(zeroIndexPage + 1);
+  }, [zeroIndexPage]);
+
+  return [zeroIndexPage, setZeroIndexPage];
+}
+
+/**
+ * Table component based on the Material React Table
+ *
+ * @see https://www.material-react-table.com/docs
+ */
+export default function Table<RowItem extends Record<string, any>>({
+  emptyMessage,
+  reflectInURL,
+  initialPage = 1,
+  rowsPerPage,
+  filterFunction,
+  errorMessage,
+  loading,
+  ...tableProps
+}: TableProps<RowItem>) {
+  const shouldReflectInURL = reflectInURL !== undefined && reflectInURL !== false;
+  const prefix = reflectInURL === true ? '' : reflectInURL || '';
+  const [page, setPage] = usePageURLState(shouldReflectInURL ? 'p' : '', prefix, initialPage);
+
+  const storeRowsPerPageOptions = useSettings('tableRowsPerPageOptions');
+  const rowsPerPageOptions = rowsPerPage || storeRowsPerPageOptions;
+  const defaultRowsPerPage = useMemo(() => helpers.getTablesRowsPerPage(rowsPerPageOptions[0]), []);
+  const [pageSize, setPageSize] = useURLState(shouldReflectInURL ? 'perPage' : '', {
+    defaultValue: defaultRowsPerPage,
+    prefix,
+  });
+
+  const { t } = useTranslation();
+  const theme = useTheme();
+
+  // Provide defaults for the columns
+  const tableColumns: TableColumn<RowItem>[] = useMemo(
+    () =>
+      tableProps.columns.map((column, i) => ({
+        ...column,
+        id: column.id ?? String(i),
+        header: column.header || '',
+      })),
+    [tableProps.columns]
+  );
+
+  const tableData = useMemo(() => {
+    if (!filterFunction) return tableProps.data ?? [];
+    return (tableProps.data ?? []).filter(it => filterFunction(it));
+  }, [tableProps.data, filterFunction]);
+
+  const gridTemplateColumns = tableProps.columns
+    .filter(it => {
+      const isHidden = tableProps.state?.columnVisibility?.[it.id!] === false;
+      return !isHidden;
+    })
+    .map(it => {
+      if (typeof it.gridTemplate === 'number') {
+        return `${it.gridTemplate}fr`;
+      }
+      return it.gridTemplate ?? '1fr';
+    })
+    .join(' ');
+
+  const table = useMaterialReactTable({
+    ...tableProps,
+    columns: tableColumns ?? [],
```

Comment:
I haven't tested this, but I don't see i18n/localization hooked up? Does anything need to be done for that?

https://www.material-react-table.com/docs/guides/localization#localization-(i18n)-guide

Reply:
added localization